zedenv/lib/system.py

```diff
@@ -77,10 +78,19 @@ def zfs_manual_mount(dataset: str, mountpoint: str, call_args: Optional[list] =
 
     if call_args:
         mount_call.extend(call_args)
+        mount_call_fallback.extend(call_args)
 
     mount_call.extend([dataset, mountpoint])
+    mount_call_fallback.extend([dataset, mountpoint])
 
     try:
         mount(call_args=mount_call)
-    except RuntimeError as e:
-        raise
+    except RuntimeError:
+        if len(mount_call) != len(mount_call_fallback):
+            try:
+                # Using `-o zfsutil` fails if we want to mount a dataset that is already mounted
```

Comment:
When would we want to mount a dataset that's already mounted? Is this necessary?

Reply:
Lets assume that our active boot environment is called `XXX`. This means, that `bpool/boot/env/zedenv-XXX` is mounted at `/boot`.

To generate the grub entries, we have to create a temporary directory structure. This requires us to remount `bpool/boot/env/zedenv-XXX` to `/boot/zfsenv/zedenv-XXX`.

Reply:
Wouldn't this only occur if the existing the environment is already in use? In which case it wouldn't be able to be mounted anyway?

https://github.com/johnramsden/zedenv/blob/9f033adf5332717b70be733d127e13b0676b4140/zedenv/cli/mount.py#L133-L151

Reply:
Actually, it is possible to mount a dataset multiple times. When I halt the execution after `setup_boot_env_tree(self)`, the following mounts are present:

     ❯ findmnt | grep boot
    ├─/boot                                     bpool/boot/env/zedenv-ubuntu  zfs         rw,relatime,xattr,posixacl
    │ ├─/boot/grub                              bpool/boot/grub               zfs         rw,nodev,relatime,xattr,posixacl
    │ ├─/boot/zfsenv/zedenv-ubuntu              bpool/boot/env/zedenv-ubuntu  zfs         rw,relatime,xattr,posixacl
    │ └─/boot/zfsenv/zedenv-default             bpool/boot/env/zedenv-default zfs         rw,relatime,xattr,posixacl

I think you are looking at the wrong place. We have to mount the boot pools each time a new `grub.cfg` must be generated. This is true for `activate`, `create`, `destroy` and `rename`.
In each case, this the call stack:

    zedenv_...
    ❯ bootloader_plugin.post_create()
      ❯ setup_boot_env_tree()
        ❯ zedenv_mount(...)
          ❯ zfs_manaual_mount(...)
    
In contrast to this, the code section you provided is the `def cli(...)` entry block which will be executed if we call `zedenv mount`.

I think this misunderstanding shows, that it could be a good idea to define some keywords at the beginning of the guide. For example:

- a ZFS boot environment (everything which is unique for each boot environment. Until now, this was the same as `root` environment of a BE)
- `root` dataset of a BE (the dataset which will be mounted to `/`)
- `boot` dataset of a BE (the dataset which will be mounted to `/boot`)

Reply:
My only concern would be whether or not that is safe. What I was doing before in the grub plug-in was doing as you explain treating the current boot environment differently. The zfsutils mount is a [bit of a hack](https://github.com/zfsonlinux/zfs/issues/7452), and I don't know if it's safe. Otherwise why is it not possible to mount a dataset twice with `zfs` or `mount -t zfs`? I think I would prefer to keep the existing behavior of treating the current `/boot` differently, that way we're not mounting a dataset more than once.

Unless I'm mistaken `zedenv.cli.mount` is what you're referring to, and it is [being used](https://github.com/johnramsden/zedenv-grub/blob/e3122eca0337d6ea6485a55366a26102255edf4a/zedenv_grub/grub.py#L250-L252), however I guess the occurrence where you're trying to mount the root dataset wouldn't happen because you're using the secondary pool, so the issue I brought up above wouldn't occur.